src/function/table/table_scan.cpp

```diff
@@ -402,10 +388,31 @@ unique_ptr<vector<unique_ptr<Expression>>> ExtractFilters(const ColumnDefinition
                                                           idx_t storage_idx) {
 	ColumnBinding binding(0, storage_idx);
 	auto bound_ref = make_uniq<BoundColumnRefExpression>(col.Name(), col.Type(), binding);
-
 	auto filter_expressions = make_uniq<vector<unique_ptr<Expression>>>();
-	ExtractInFilter(filter, *bound_ref, filter_expressions);
 
+	switch (filter->filter_type) {
+	case TableFilterType::OPTIONAL_FILTER:
+		// Conventional IN filter.
+		ExtractOptionalFilter(*filter, *bound_ref, filter_expressions, false);
+		break;
+	case TableFilterType::CONJUNCTION_AND: {
+		// Dynamic filter pushed down via a hash join.
+		auto &and_filter = filter->Cast<ConjunctionAndFilter>();
+		auto &children = and_filter.child_filters;
+		if (children.empty()) {
+			break;
+		}
+		if (children[0]->filter_type != TableFilterType::OPTIONAL_FILTER) {
+			break;
+		}
+		ExtractOptionalFilter(*children[0], *bound_ref, filter_expressions, true);
```

Comment:
Let's try to remove the `origin_is_hash_join` entirely and do proper pruning of the `IN` list based on the other filters